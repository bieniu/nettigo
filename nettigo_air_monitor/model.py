"""Type definitions for NAM."""
from dataclasses import dataclass
from typing import Optional


@dataclass
class NAMSensors:  # pylint: disable=too-many-instance-attributes
    """Data class for NAM sensors."""

    bme280_humidity: Optional[float]
    bme280_pressure: Optional[float]
    bme280_temperature: Optional[float]
    bmp280_pressure: Optional[float]
    bmp280_temperature: Optional[float]
    dht22_humidity: Optional[float]
    dht22_temperature: Optional[float]
    heca_humidity: Optional[float]
    heca_temperature: Optional[float]
    mhz14a_carbon_dioxide: Optional[int]
    sds_p1: Optional[float]
    sds_p2: Optional[float]
    sht3x_humidity: Optional[float]
    sht3x_temperature: Optional[float]
    signal: Optional[int]
    sps30_p0: Optional[float]
    sps30_p1: Optional[float]
    sps30_p2: Optional[float]
    sps30_p4: Optional[float]
    uptime: Optional[int]
