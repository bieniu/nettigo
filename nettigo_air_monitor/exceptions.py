"""NAM exceptions."""
from __future__ import annotations


class NamError(Exception):
    """Base class for NAM errors."""

    def __init__(self, status: str) -> None:
        """Initialize."""
        super().__init__(status)
        self.status = status


class ApiError(NamError):
    """Raised when request ended in error."""


class AuthRequired(NamError):
    """Raised if auth is required but not given."""


class CannotGetMac(NamError):
    """Raised when cannot get device MAC address."""


class InvalidSensorData(NamError):
    """Raised when sensor data is invalid."""
