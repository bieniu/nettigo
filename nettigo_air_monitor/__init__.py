"""Python wrapper for getting air quality data from Nettigo Air Monitor devices."""
from __future__ import annotations

import asyncio
import logging
import re
from http import HTTPStatus
from typing import Any, cast

from aiohttp import ClientSession
from aiohttp.client_exceptions import ClientConnectorError
from dacite import from_dict

from .const import (
    ATTR_DATA,
    ATTR_UPTIME,
    ATTR_VALUES,
    ENDPOINTS,
    MAC_PATTERN,
    RENAME_KEY_MAP,
    RETRIES,
    TIMEOUT,
)
from .exceptions import ApiError, AuthRequired, CannotGetMac, InvalidSensorData
from .model import ConnectionOptions, NAMSensors

_LOGGER = logging.getLogger(__name__)


async def create_device(session: ClientSession, options: ConnectionOptions):
    """Create a new device instance."""
    device = NettigoAirMonitor(session, options)
    await device.initialize()
    return device


class NettigoAirMonitor:
    """Main class to perform Nettigo Air Monitor requests."""

    def __init__(self, session: ClientSession, options: ConnectionOptions) -> None:
        """Initialize."""
        self._session = session
        self._host = options.host
        self._options = options
        self._software_version: str

    async def initialize(self) -> None:
        """Initialize."""
        _LOGGER.debug("Initializing device %s", self._host)

        if not await self._async_post_data("config"):
            raise AuthRequired("(Re)auth is required")

    @staticmethod
    def _construct_url(arg: str, **kwargs: str) -> str:
        """Construct Nettigo Air Monitor URL."""
        return ENDPOINTS[arg].format(**kwargs)

    @staticmethod
    def _parse_sensor_data(data: dict[Any, Any]) -> dict[str, int | float]:
        """Parse sensor data dict."""
        result = {
            item["value_type"].lower(): round(float(item["value"]), 1) for item in data
        }

        for key, value in result.items():
            if "pressure" in key and value is not None:
                result[key] = round(value / 100)
            if (
                key
                in (
                    "conc_co2_ppm",
                    "sds_p1",
                    "sds_p2",
                    "sps30_p0",
                    "sps30_p1",
                    "sps30_p2",
                    "sps30_p4",
                    "signal",
                )
                and value is not None
            ):
                result[key] = round(value)

        for old_key, new_key in RENAME_KEY_MAP:
            if result.get(old_key) is not None:
                result[new_key] = result.pop(old_key)

        return result

    async def _async_get_data(self, url: str, use_json: bool = True) -> Any:
        """Retrieve data from the device."""
        last_error = None
        for retry in range(RETRIES):
            try:
                resp = await self._session.get(url)
            except ClientConnectorError as error:
                _LOGGER.info(
                    "Invalid response from device: %s, retry: %s", self._host, retry
                )
                last_error = error
            else:
                _LOGGER.debug(
                    "Data retrieved from %s, status: %s", self._host, resp.status
                )
                if resp.status != HTTPStatus.OK.value:
                    raise ApiError(
                        f"Invalid response from device {self._host}: {resp.status}"
                    )

                return await resp.json() if use_json else await resp.text()

            wait = TIMEOUT + retry
            _LOGGER.debug("Waiting %s seconds for device %s", wait, self._host)
            await asyncio.sleep(wait)

        raise ApiError(str(last_error))

    async def async_update(self) -> NAMSensors:
        """Retrieve data from the device."""
        url = self._construct_url(ATTR_DATA, host=self._host)

        data = await self._async_get_data(url)

        self._software_version = data["software_version"]

        try:
            sensors = self._parse_sensor_data(data["sensordatavalues"])
        except (TypeError, KeyError) as err:
            raise InvalidSensorData("Invalid sensor data") from err

        if ATTR_UPTIME in data:
            sensors[ATTR_UPTIME] = int(data[ATTR_UPTIME])

        return from_dict(data_class=NAMSensors, data=sensors)

    async def async_get_mac_address(self) -> str:
        """Retrieve the device MAC address."""
        url = self._construct_url(ATTR_VALUES, host=self._host)
        data = await self._async_get_data(url, use_json=False)

        if not (mac := re.search(MAC_PATTERN, data)):
            raise CannotGetMac("Cannot get MAC address from device")

        return cast(str, mac[0])

    @property
    def software_version(self) -> str:
        """Return software version."""
        return self._software_version

    async def _async_post_data(self, path: str) -> bool:
        """Perform a HTTP post request."""
        url = url = f"http://{self._host}/{path}"
        resp = await self._session.post(url, auth=self._options.auth, timeout=TIMEOUT)

        return resp.status == HTTPStatus.OK.value

    async def restart(self) -> bool:
        """Restart the device."""
        return await self._async_post_data("reset")

    async def ota_update(self) -> bool:
        """Trigger OTA update."""
        return await self._async_post_data("ota")
