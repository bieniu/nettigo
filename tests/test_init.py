"""Tests for nettigo package."""
import json
from unittest.mock import Mock

import aiohttp
import pytest
from aiohttp.client_exceptions import ClientConnectorError
from aioresponses import aioresponses

from nettigo_air_monitor import (
    ApiError,
    CannotGetMac,
    InvalidSensorData,
    NettigoAirMonitor,
)

VALID_IP = "192.168.172.12"
INVALID_HOST = "http://nam.org"

VALUES = "MAC: AA:BB:CC:DD:EE:FF<br/>"


@pytest.mark.asyncio
async def test_valid_data():
    """Test with valid data."""
    with open("tests/fixtures/valid_data.json", encoding="utf-8") as file:
        data = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://192.168.172.12/data.json",
            payload=data,
        )
        session_mock.get(
            "http://192.168.172.12/values",
            payload=VALUES,
        )

        nam = NettigoAirMonitor(session, VALID_IP)

        result = await nam.async_get_mac_address()

        assert result == "AA:BB:CC:DD:EE:FF"

        result = await nam.async_update()

    await session.close()

    assert nam.software_version == "NAMF-2020-36"
    assert result.bme280_humidity == 85.3
    assert result.bme280_pressure == 989
    assert result.bme280_temperature == 10.6
    assert result.bmp280_pressure == 1022
    assert result.bmp280_temperature == 5.6
    assert result.dht22_humidity == 46.2
    assert result.dht22_temperature == 6.3
    assert result.heca_humidity == 59.7
    assert result.heca_temperature == 15.1
    assert result.mhz14a_carbon_dioxide == 865
    assert result.sds011_p1 == 23
    assert result.sds011_p2 == 20
    assert result.sht3x_humidity == 34.7
    assert result.sht3x_temperature == 6.3
    assert result.signal == -85
    assert result.sps30_p0 == 31
    assert result.sps30_p1 == 21
    assert result.sps30_p2 == 34
    assert result.sps30_p4 == 25
    assert result.uptime == 45632


@pytest.mark.asyncio
async def test_api_error():
    """Test API error."""
    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://192.168.172.12/data.json",
            status=404,
        )

        nam = NettigoAirMonitor(session, VALID_IP)

        try:
            await nam.async_update()
        except ApiError as error:
            assert (
                str(error.status) == "Invalid response from device 192.168.172.12: 404"
            )

    await session.close()


@pytest.mark.asyncio
async def test_retry():
    """Test retry request."""
    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://192.168.172.12/data.json",
            exception=ClientConnectorError(Mock(), Mock()),
        )
        session_mock.get(
            "http://192.168.172.12/data.json",
            exception=ClientConnectorError(Mock(), Mock()),
        )
        session_mock.get(
            "http://192.168.172.12/data.json",
            exception=ClientConnectorError(Mock(), Mock()),
        )
        session_mock.get(
            "http://192.168.172.12/data.json",
            exception=ClientConnectorError(Mock(), Mock()),
        )

        nam = NettigoAirMonitor(session, VALID_IP)

        try:
            await nam.async_update()
        except ApiError as error:
            assert "Cannot connect to host" in str(error)

    await session.close()


@pytest.mark.asyncio
async def test_invalid_sensor_data():
    """Test InvalidSensorData error."""
    with open("tests/fixtures/invalid_data.json", encoding="utf-8") as file:
        data = json.load(file)

    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://192.168.172.12/data.json",
            payload=data,
        )
        nam = NettigoAirMonitor(session, VALID_IP)

        try:
            await nam.async_update()
        except InvalidSensorData as error:
            assert str(error.status) == "Invalid sensor data"

    await session.close()


@pytest.mark.asyncio
async def test_cannot_get_mac():
    """Test CannotGetMac error."""
    session = aiohttp.ClientSession()

    with aioresponses() as session_mock:
        session_mock.get(
            "http://192.168.172.12/values",
            payload="lorem ipsum",
        )
        nam = NettigoAirMonitor(session, VALID_IP)

        try:
            await nam.async_get_mac_address()
        except CannotGetMac as error:
            assert str(error.status) == "Cannot get MAC address from device"

    await session.close()
