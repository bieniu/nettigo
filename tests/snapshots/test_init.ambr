# serializer version: 1
# name: test_caqi_value
  NAMSensors(bme280_humidity=None, bme280_pressure=None, bme280_temperature=None, bmp180_pressure=None, bmp180_temperature=None, bmp280_pressure=None, bmp280_temperature=None, dht22_humidity=None, dht22_temperature=None, ds18b20_temperature=None, gps_altitude=None, gps_latitude=None, gps_longitude=None, heca_humidity=None, heca_temperature=None, mhz14a_carbon_dioxide=None, pms_caqi=None, pms_caqi_level=None, pms_p0=None, pms_p1=None, pms_p2=None, sds011_caqi=None, sds011_caqi_level=None, sds011_p1=None, sds011_p2=None, sht3x_humidity=None, sht3x_temperature=None, signal=None, sps30_caqi=None, sps30_caqi_level=None, sps30_p0=None, sps30_p1=None, sps30_p2=None, sps30_p4=None, uptime=None)
# ---
# name: test_data_cached
  NAMSensors(bme280_humidity=85.3, bme280_pressure=989.206, bme280_temperature=10.6, bmp180_pressure=996.784, bmp180_temperature=10.8, bmp280_pressure=1022.012, bmp280_temperature=5.6, dht22_humidity=46.2, dht22_temperature=6.3, heca_humidity=59.7, heca_temperature=15.1, mhz14a_carbon_dioxide=865.0, pms_caqi=19, pms_caqi_level='very_low', pms_p0=6.0, pms_p1=10.0, pms_p2=11.0, sds011_caqi=34, sds011_caqi_level='low', sds011_p1=22.7, sds011_p2=20.0, sht3x_humidity=34.7, sht3x_temperature=6.3, signal=-85.0, sps30_caqi=54, sps30_caqi_level='medium', sps30_p0=31.2, sps30_p1=21.2, sps30_p2=34.3, sps30_p4=24.7, uptime=45632)
# ---
# name: test_valid_data
  NAMSensors(bme280_humidity=85.3, bme280_pressure=989.206, bme280_temperature=10.6, bmp180_pressure=996.784, bmp180_temperature=10.8, bmp280_pressure=1022.012, bmp280_temperature=5.6, dht22_humidity=46.2, dht22_temperature=6.3, ds18b20_temperature=12.9, gps_altitude=102.8, gps_latitude=52.3, gps_longitude=20.9, heca_humidity=59.7, heca_temperature=15.1, mhz14a_carbon_dioxide=865.0, pms_caqi=19, pms_caqi_level='very_low', pms_p0=6.0, pms_p1=10.0, pms_p2=11.0, sds011_caqi=34, sds011_caqi_level='low', sds011_p1=22.7, sds011_p2=20.0, sht3x_humidity=34.7, sht3x_temperature=6.3, signal=-85.0, sps30_caqi=54, sps30_caqi_level='medium', sps30_p0=31.2, sps30_p1=21.2, sps30_p2=34.3, sps30_p4=24.7, uptime=45632)
# ---
# name: test_valid_data_with_auth
  NAMSensors(bme280_humidity=85.3, bme280_pressure=989.206, bme280_temperature=10.6, bmp180_pressure=996.784, bmp180_temperature=10.8, bmp280_pressure=1022.012, bmp280_temperature=5.6, dht22_humidity=46.2, dht22_temperature=6.3, ds18b20_temperature=12.9, gps_altitude=102.8, gps_latitude=52.3, gps_longitude=20.9, heca_humidity=59.7, heca_temperature=15.1, mhz14a_carbon_dioxide=865.0, pms_caqi=19, pms_caqi_level='very_low', pms_p0=6.0, pms_p1=10.0, pms_p2=11.0, sds011_caqi=34, sds011_caqi_level='low', sds011_p1=22.7, sds011_p2=20.0, sht3x_humidity=34.7, sht3x_temperature=6.3, signal=-85.0, sps30_caqi=54, sps30_caqi_level='medium', sps30_p0=31.2, sps30_p1=21.2, sps30_p2=34.3, sps30_p4=24.7, uptime=45632)
# ---
